### PR TITLE
Pulumi code for deploying a Dagster instance

### PR DESCRIPTION
**Pull request by blarghmatey**
This adds the logic and bug fixes to allow for deploying a Dagster instance from the AMI built via Packer as defined [here](https://github.com/mitodl/salt-ops/blob/main/packer/dagster.json).

### CODE REVIEW

File: src/ol_infrastructure/components/aws/database.py [hunk 1]
```diff
@@ -166,37 +168,37 @@ def __init__(self, db_config: OLDBConfig, opts: pulumi.ResourceOptions = None):
...
-            parameter_group_name=parameter_group.name,
+            parameter_group_name=self.parameter_group.name,
             password=db_config.password.get_secret_value(),
             port=db_config.port,
             publicly_accessible=db_config.is_public,
             skip_final_snapshot=False,
             storage_encrypted=True,
             storage_type=db_config.storage_type.value,
-            tags=db_config.tags,
+            # tags=db_config.tags,
```
Comment (shaidar) [inline src/ol_infrastructure/components/aws/database.py:178]: Why is this commented out?

File: src/ol_infrastructure/components/services/vault.py [hunk 2]
```diff
@@ -22,47 +22,58 @@
...
     mysql_rds = 'mysql_rds'
     mongodb = 'mongodb'
 
+
 class OLVaultDatabaseConfig(BaseModel):
     db_name: Text
     mount_point: Text
     db_admin_username: Text
-    db_admin_password: SecretStr
+    db_admin_password: Text
```
Comment (shaidar) [inline src/ol_infrastructure/components/services/vault.py:41]: Why the change here?

File: src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.Production.yaml [hunk 3]
```diff
@@ -2,3 +2,12 @@ secretsprovider: awskms://alias/PulumiSecrets
 encryptedkey: AQICAHionUR8LBW1ALuVC0rCH3AE2oQIfGMCx3XmpDH9HjM2LQEGccCx6w+aU1z9VNPzXzwIAAAAfjB8BgkqhkiG9w0BBwagbzBtAgEAMGgGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMkTCVUUXv23ZgiFtbAgEQgDusl6M2Rs0caUtV7jakSKAqsXTqm8f7M0UuEu0hjZ5KdbCp8HUDBShST/yKEpt0ExGXlODrlmL3gGN6fg==
 config:
   aws:region: us-east-1
+  dagster:instance_type: m5a.large
+  dagster:domain: pipelines-qa.odl.mit.edu
```
Comment (shaidar) [inline src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.Production.yaml:6]: Should this be `production`?

File: src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.QA.yaml [hunk 4]
```diff
@@ -4,3 +4,11 @@ config:
...
   dagster:db_password:
     secure: v1:y15TVAw5jqVNTrXB:zN/lS+0N6NeqUxTGAwY64IV5LFyXHzDRZloAZwwgQA==
+  dagster:domain: pipelines-qa.odl.mit.edu
+  dagster:instance_type: t3a.medium
+  saltstack:api_password:
+    secure: v1:jcxzqwIOoxQ2261K:PsHmBDKsQetmclqZPtQmAD43OgsnLlDu9NAVVVP9iNtNzPmFNb1fj7QeBANm1uxyhY1cMN7spWYR2w==
+  saltstack:api_url: https://salt-qa.odl.mit.edu
+  saltstack:api_user: pulumi
+  saltstack:key_name: salt-qa
+  vault:skip_tls_verify: "true"
```
Comment (shaidar) [inline src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.QA.yaml:14]: Is this missing the `vault:address`?

File: src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.Production.yaml [hunk 5]
```diff
@@ -2,3 +2,12 @@ secretsprovider: awskms://alias/PulumiSecrets
 encryptedkey: AQICAHionUR8LBW1ALuVC0rCH3AE2oQIfGMCx3XmpDH9HjM2LQEGccCx6w+aU1z9VNPzXzwIAAAAfjB8BgkqhkiG9w0BBwagbzBtAgEAMGgGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMkTCVUUXv23ZgiFtbAgEQgDusl6M2Rs0caUtV7jakSKAqsXTqm8f7M0UuEu0hjZ5KdbCp8HUDBShST/yKEpt0ExGXlODrlmL3gGN6fg==
 config:
   aws:region: us-east-1
+  dagster:instance_type: m5a.large
+  dagster:domain: pipelines.odl.mit.edu
+  dagster:db_password:
```
Comment (shaidar) [inline src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.Production.yaml:7]: Is this supposed to be `saltstack:api_password` cause that's what's in dagster.QA

File: src/ol_infrastructure/infrastructure/aws/network/Pulumi.infrastructure.aws.network.Dev.yaml [hunk 6]
```diff
@@ -1,2 +1,7 @@
 secretsprovider: awskms://alias/PulumiSecrets
 encryptedkey: AQICAHionUR8LBW1ALuVC0rCH3AE2oQIfGMCx3XmpDH9HjM2LQE5JVlRpVbYeiNKFwqxSeLFAAAAfjB8BgkqhkiG9w0BBwagbzBtAgEAMGgGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMCnctp6kjfnFHFrOCAgEQgDukpJ8nxbBFRMGqo5VuudSbWHeCV5XzBVj28cgetVPMBOlbke4WOClmQi4cYkKYfAKWgSvDmQUtAb7rUg==
+config:
+  aws:region: us-east-1
+  data_vpc:cidr_block: 10.200.0.0/16
+  operations_vpc:cidr_block: 10.201.0.0/16
```
Comment (shaidar) [inline src/ol_infrastructure/infrastructure/aws/network/Pulumi.infrastructure.aws.network.Dev.yaml:6]: Should there be a `data_vpc:name`?

File: src/ol_infrastructure/components/services/vault.py (hunk 2 above)
Comment (blarghmatey) [reply src/ol_infrastructure/components/services/vault.py:41]: Because it was adding unnecessary complexity and I just wanted to remove it as a possible source of error.

File: src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.QA.yaml (hunk 4 above)
Comment (blarghmatey) [reply src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.QA.yaml:14]: That can be supplied by an environment variable. I need to add some details to the readme for that.

File: src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.Production.yaml (hunk 5 above)
Comment (blarghmatey) [reply src/ol_infrastructure/infrastructure/applications/dagster/Pulumi.applications.dagster.Production.yaml:7]: No, it's for the db password. I haven't added the salt api password because we haven't deployed those changes to production yet.

File: src/ol_infrastructure/infrastructure/aws/network/Pulumi.infrastructure.aws.network.Dev.yaml (hunk 6 above)
Comment (blarghmatey) [reply src/ol_infrastructure/infrastructure/aws/network/Pulumi.infrastructure.aws.network.Dev.yaml:6]: I added the operations VPC name because of the fact that it already exists and we need to import it.

File: src/ol_infrastructure/components/aws/olvpc.py [hunk 7]
```diff
@@ -234,7 +229,7 @@ def __init__(
         :type opts: ResourceOptions
         """
         super().__init__('ol:infrastructure:aws:VPCPeeringConnection', vpc_peer_name, None, opts)
-        resource_options = ResourceOptions(parent=self).merge(opts)
+        resource_options = ResourceOptions.merge(ResourceOptions(parent=self), opts)  # type: ignore
```
Comment (shaidar) [inline src/ol_infrastructure/components/aws/olvpc.py:232]: Don't think this is making sense to me

File: src/ol_infrastructure/components/services/vault.py [hunk 8]
```diff
@@ -22,47 +22,58 @@
...
     db_type: Text = DBEngines.postgres.value
-    role_statements: Dict[Text, Text] = postgresql_sql_statements
+    role_statements: Dict[Text, Dict[Text, Text]] = postgres_sql_statements
 
 
 class OLVaultMysqlDatabaseConfig(OLVaultDatabaseConfig):
-    db_port: Int = DEFAULT_PORT_MYSQL
-    db_connection: Text = '{{db_user}}:{{db_pass}}@tcp({{db_host}}:{{db_port}})/'
+    db_port: int = DEFAULT_PORT_MYSQL
+    db_connection: Text = '{{{{username}}}}:{{{{password}}}}@tcp({db_host}:{db_port})/'
```
Comment (shaidar) [inline src/ol_infrastructure/components/services/vault.py:62]: Where are the `username` and `password` defined?

Comment (blarghmatey) [reply src/ol_infrastructure/components/services/vault.py:62]: Those are getting templated in from the `data` block that gets passed to the database secret backend class. It also allows for rotating the root credentials within Vault so that there's no way to recover them exernally.

File: src/ol_infrastructure/components/aws/olvpc.py (hunk 7 above)
Comment (blarghmatey) [reply src/ol_infrastructure/components/aws/olvpc.py:232]: That was just attempting to address a MyPy complaint. The two versions are functionally equivalent.

File: src/ol_infrastructure/infrastructure/applications/dagster/__main__.py [hunk 9]
```diff
@@ -110,26 +119,126 @@
...
     tags=aws_config.tags,
     db_name='dagster',
     **defaults(stack)['rds'],
 )
 dagster_db = OLAmazonDB(dagster_db_config)
+
+dagster_db_vault_backend_config = OLVaultPostgresDatabaseConfig(
+    db_name=dagster_db_config.db_name,
+    mount_point=f'{dagster_db_config.engine}-dagster-{dagster_environment}',
+    db_admin_username=dagster_db_config.username,
```
Comment (shaidar) [inline src/ol_infrastructure/infrastructure/applications/dagster/__main__.py:147]: Where is the `dagster_db_config.username` defined as I don't see it in dagster_db_config

Comment (blarghmatey) [reply src/ol_infrastructure/infrastructure/applications/dagster/__main__.py:147]: It's coming from the default value defined in the OLDatabaseConfig base class here https://github.com/mitodl/ol-infrastructure/blob/dagster_instance/src/ol_infrastructure/components/aws/database.py#L71

File: src/ol_infrastructure/lib/vault.py [hunk 10]
```diff
@@ -1,72 +1,87 @@
 # noqa: WPS226
+'''Creation and revocation statements used for Vault role definitions.'''
 
+# These strings are passed through the `.format` method so the variables that need to remain in the template
+# to be passed to Vault are wrapped in 4 pairs of braces. TMM 2020-09-01
 postgres_sql_statements = {
+    'approle': {
+        'create': 'CREATE ROLE {role_name};',
+        'revoke': 'DROP ROLE {role_name}'
```
Comment (shaidar) [inline src/ol_infrastructure/lib/vault.py:9]: Missing `;`?

**Comment by shaidar** [comment]
👍
